File: backend/app/schemas/records.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class FeedingRecordCreate(BaseModel):
    babyId: int = Field(gt=0)
    occurredAt: datetime
    feedingType: Literal["BREAST", "FORMULA", "MIXED"]
    amountMl: int | None = Field(default=None, gt=0, le=500)
    durationMinutes: int | None = Field(default=None, gt=0, le=180)
    breastSide: Literal["LEFT", "RIGHT", "BOTH"] | None = None
    burped: bool | None = None
    memo: str | None = Field(default=None, max_length=1000)

    @model_validator(mode="after")
    def validate_amount_or_duration(self):
        if self.amountMl is None and self.durationMinutes is None:
            raise ValueError("amountMl or durationMinutes is required")
        return self


class SleepRecordCreate(BaseModel):
    babyId: int = Field(gt=0)
    startedAt: datetime
    endedAt: datetime
    sleepType: Literal["NAP", "NIGHT"] | None = None
    status: Literal["PEACEFUL", "RESTLESS", "WOKE_OFTEN"] | None = None

    @model_validator(mode="after")
    def validate_period(self):
        if self.endedAt <= self.startedAt:
            raise ValueError("endedAt must be after startedAt")
        return self


class UrineRecordCreate(BaseModel):
    babyId: int = Field(gt=0)
    occurredAt: datetime
    amount: Literal["SMALL", "MEDIUM", "LARGE"] | None = None
    color: Literal["NORMAL", "DARK_YELLOW", "PINK", "RED", "OTHER"] | None = None


class StoolRecordCreate(BaseModel):
    babyId: int = Field(gt=0)
    occurredAt: datetime
    amount: Literal["SMALL", "MEDIUM", "LARGE"] | None = None
    color: Literal["NORMAL", "GREEN", "BLACK", "RED", "WHITE", "OTHER"] | None = None
    form: Literal["WATERY", "SOFT", "NORMAL", "HARD"] | None = None
    photoUrl: str | None = Field(default=None, max_length=1000)

File: frontend/src/features/records/types/records.ts
export type RecordType = "FEEDING" | "SLEEP" | "URINE" | "STOOL";

export type RecordContent = {
  amount?: "SMALL" | "MEDIUM" | "LARGE" | null;
  breastSide?: "LEFT" | "RIGHT" | "BOTH" | null;
  burped?: boolean | null;
  color?: string | null;
  durationMinutes?: number | null;
  feedingType?: "BREAST" | "FORMULA" | "MIXED";
  form?: "WATERY" | "SOFT" | "NORMAL" | "HARD" | null;
  formulaAmountMl?: number;
  photoUrl?: string | null;
  sleepType?: "NAP" | "NIGHT" | null;
  status?: "PEACEFUL" | "RESTLESS" | "WOKE_OFTEN" | null;
};

export type CareRecord = {
  id: string;
  type: RecordType;
  occurredAt: string;
  startedAt: string | null;
  endedAt: string | null;
  content: RecordContent;
  memo: string | null;
  createdAt: string;
};

type BaseCreateBody = { babyId: number };

export type FeedingRecordCreate = BaseCreateBody & {
  occurredAt: string;
  feedingType: "BREAST" | "FORMULA" | "MIXED";
  amountMl?: number;
  durationMinutes?: number;
  breastSide?: "LEFT" | "RIGHT" | "BOTH";
  burped?: boolean;
  memo?: string;
};

export type SleepRecordCreate = BaseCreateBody & {
  startedAt: string;
  endedAt: string;
  sleepType?: "NAP" | "NIGHT";
  status?: "PEACEFUL" | "RESTLESS" | "WOKE_OFTEN";
};

export type UrineRecordCreate = BaseCreateBody & {
  occurredAt: string;
  amount?: "SMALL" | "MEDIUM" | "LARGE";
  color?: "NORMAL" | "DARK_YELLOW" | "PINK" | "RED" | "OTHER";
};

export type StoolRecordCreate = BaseCreateBody & {
  occurredAt: string;
  amount?: "SMALL" | "MEDIUM" | "LARGE";
  color?: "NORMAL" | "GREEN" | "BLACK" | "RED" | "WHITE" | "OTHER";
  form?: "WATERY" | "SOFT" | "NORMAL" | "HARD";
  photoUrl?: string;
};

export type RecordCreateBody = FeedingRecordCreate | SleepRecordCreate | UrineRecordCreate | StoolRecordCreate;

export type RecordListQuery = {
  babyId: number;
  date: string;
  type?: RecordType;
  cursor?: string;
  limit?: number;
};

export type RecordListResult = {
  records: CareRecord[];
  cursor: string | null;
  hasNext: boolean;
};
